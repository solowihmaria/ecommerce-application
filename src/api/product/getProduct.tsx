import axios from 'axios';
import type { ProductResponse } from './product.types';
import type { AuthResponse } from '../createCustomer/createCustomer.types';
import { getToken } from '../token';

export const getUserToken = async (loginStatus: boolean) => {
    let accessToken: string = '';

    if (loginStatus) {
        const storedToken = getToken();
        if (storedToken) {
            accessToken = storedToken;
        }
    } else {
        const tokenResponse = await getGuestToken();
        accessToken = tokenResponse.access_token;
    }

    return accessToken;
};

const getGuestToken = async () => {
    const authUrl = process.env.CTP_AUTH_URL;
    const clientId = process.env.CTP_CLIENT_ID;
    const clientSecret = process.env.CTP_CLIENT_SECRET;

    const tokenUrl = `${authUrl}/oauth/token`;
    const credentials = btoa(`${clientId}:${clientSecret}`);

    const response = await axios.post<AuthResponse>(
        tokenUrl,
        new URLSearchParams({
            grant_type: 'client_credentials',
        }),
        {
            headers: {
                Authorization: `Basic ${credentials}`,
                'Content-type': 'application/x-www-form-urlencoded',
            },
        }
    );

    return response.data;
};

export const getProductByID = async (
    id: string,
    token: string
): Promise<ProductResponse> => {
    const apiUrl = process.env.CTP_API_URL;
    const projectKey = process.env.CTP_PROJECT_KEY;

    const productUrl = `${apiUrl}/${projectKey}/products/${id}`;
    const product = await axios.get<ProductResponse>(productUrl, {
        headers: {
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
    });
    return product.data;
};
