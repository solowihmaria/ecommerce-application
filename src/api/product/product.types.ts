export enum Attributes {
    family = 'family',
    size = 'size',
    careLevel = 'care-level',
    toxic = 'toxic',
    lightRequirements = 'light-requirements',
    height = 'height',
}

export enum Sizes {
    small = 'S',
    medium = 'M',
    large = 'L',
}

export enum Care {
    easy = 'easy',
    medium = 'medium',
    hard = 'hard',
}

export enum Light {
    low = 'low',
    medium = 'medium',
    high = 'high',
}

export interface ProductResponse {
    id: string;
    masterData: {
        current: {
            name: {
                'en-US': string;
            };
            description: {
                'en-US': string;
            };
            slug: {
                'en-US': string;
            };
            masterVariant: Variant;
            variants: Variant[];
        };
    };
}

export interface Variant {
    id: number;
    sku: string;
    prices: Price[];
    images: Image[];
    attributes: (
        | FamilyAttribute
        | SizeAttribute
        | CareAttribute
        | ToxicAttribute
        | LightAttribute
        | HeightAttribute
    )[];
}

export type AttributesList = (
    | [string, string]
    | [string, Sizes]
    | [string, Care]
    | [string, boolean]
    | [string, Light]
    | [string, number]
)[];

export interface Price {
    id: string;
    value: {
        type: string;
        currencyCode: string;
        centAmount: number;
        fractionDigits: number;
    };
    discounted?: {
        value: {
            type: string;
            currencyCode: string;
            centAmount: number;
            fractionDigits: number;
        };
    };
    key: string;
}

export interface Image {
    url: string;
    label: string;
    dimensions: {
        w: number;
        h: number;
    };
}

interface FamilyAttribute {
    name: Attributes.family;
    value: string;
}

interface SizeAttribute {
    name: Attributes.size;
    value: {
        key: Sizes;
        label: Sizes;
    };
}

interface CareAttribute {
    name: Attributes.careLevel;
    value: {
        key: Care;
        label: Care;
    };
}

interface ToxicAttribute {
    name: Attributes.toxic;
    value: boolean;
}

interface HeightAttribute {
    name: Attributes.height;
    value: number;
}

interface LightAttribute {
    name: Attributes.lightRequirements;
    value: {
        key: Light;
        label: Light;
    };
}

export interface CustomProduct {
    id: string;
    name: string;
    description: string;
    masterVariant: CustomVariant;
    variants: CustomVariant[];
}

export interface CustomVariant {
    id: number;
    sku: string;
    price: number;
    images: Image[];
    family: string;
    size: Sizes;
    care: Care;
    toxic: boolean;
    height: number;
    light: Light;
    discount: number | null;
}

export interface CustomAttributes {
    family: string;
    size: Sizes;
    'care-level': Care;
    toxic: boolean;
    height: number;
    'light-requirements': Light;
}
