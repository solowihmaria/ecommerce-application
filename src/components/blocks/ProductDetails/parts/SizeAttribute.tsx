import clsx from 'clsx';
import type {
    CustomProduct,
    CustomVariant,
} from '../../../../api/product/product.types';
import styles from '../ProductDetails.module.scss';

export const SizeAttribute = ({
    product,
    currentProductVariant,
    onSizeChange,
}: {
    product: CustomProduct;
    currentProductVariant: CustomVariant;
    onSizeChange: (size: string) => void;
}) => {
    const getAvailableSizes = (product: CustomProduct) => {
        const sizes: string[] = [];
        sizes.push(product.masterVariant.size);
        product.variants.forEach((variant) => {
            sizes.push(variant.size);
        });
        return sizes;
    };
    const getCurrentSize = (currentProductVariant: CustomVariant) => {
        return currentProductVariant.size;
    };

    const getSizeName = (size: string) => {
        let sizeFullName = '';
        switch (size) {
            case 'S':
                sizeFullName = 'Small';
                break;
            case 'M':
                sizeFullName = 'Medium';
                break;
            case 'L':
                sizeFullName = 'Large';
                break;
        }
        return sizeFullName;
    };

    return (
        <ul className={styles.sizeContainer}>
            {getAvailableSizes(product).map((size) =>
                size === getCurrentSize(currentProductVariant) ? (
                    <li
                        key={size}
                        className={clsx(styles.sizeActive, styles.size)}
                    >
                        {getSizeName(size)}
                    </li>
                ) : (
                    <li
                        onClick={() => {
                            console.log(`hello size ${size}`);
                            onSizeChange(size);
                        }}
                        aria-hidden="true"
                        key={size}
                        className={styles.size}
                    >
                        {getSizeName(size)}
                    </li>
                )
            )}
        </ul>
    );
};
