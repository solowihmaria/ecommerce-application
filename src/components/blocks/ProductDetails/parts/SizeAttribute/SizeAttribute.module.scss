@use '../../../../../assets/styles/constants.module' as c;

.sizeContainer {
    display: flex;
    flex-wrap: wrap;
    gap: 15px;
    justify-content: flex-start;
    list-style-type: none;
}

.button.size {
    background: none;
}

.size.sizeActive {
    color: c.$color-accent-dark;
    background-color: c.$color-gray-300;
}

.size.sizeActive:hover {
    color: c.$color-text-primary;
}

@media (width <= 992px) {
    .sizeContainer {
        gap: 1rem;
        justify-content: flex-start;
    }

    .button.size {
        font-size: 1rem;
    }
}

@media (width <= 500px) {
    .sizeContainer {
        justify-content: flex-start;
    }

    .button.size {
        padding: 0.5rem 1rem;
        font-size: 0.875rem;
    }
}
