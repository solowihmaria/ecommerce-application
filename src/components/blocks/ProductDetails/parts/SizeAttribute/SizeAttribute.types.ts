import type { CustomProduct } from '../../../../../api/product/product.types';

import type { CustomVariant, Sizes } from '../../../../../types/product.types';

export interface SizeAttributeProps {
    product: CustomProduct;
    currentProductVariant: CustomVariant;
    onSizeChange: (size: Sizes) => void;
}
