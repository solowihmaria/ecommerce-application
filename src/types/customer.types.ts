export enum CountryCode {
    FR = 'FR',
    DE = 'DE',
    IT = 'IT',
}

export type AddressType = 'shipping' | 'billing';

export interface Address {
    id?: string;
    key?: string;
    streetName: string;
    postalCode: string;
    city: string;
    country: CountryCode;
    type?: AddressType;
    isDefault?: boolean;
}

export interface BaseCustomer {
    email: string;
    firstName: string;
    lastName: string;
    dateOfBirth?: string; // Опционально в профиле, но обязательно при регистрации
    addresses: Address[];
    defaultShippingAddressId?: string | null;
    defaultBillingAddressId?: string | null;
}
