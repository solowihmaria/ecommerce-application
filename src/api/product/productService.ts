import { getProductByID } from './getProduct';
import type {
    AttributesList,
    Care,
    CustomAttributes,
    CustomProduct,
    CustomVariant,
    Light,
    ProductResponse,
    Sizes,
    Variant,
} from './product.types';

export const getProductData = async (id: string) => {
    const productData = await getProductByID(id);

    return productData;
};

const prepareVariant = (initialVariant: Variant): CustomVariant => {
    const { id, sku, images, prices } = initialVariant;
    //filter out not used attributes
    const attributes = initialVariant.attributes.filter((attribute) =>
        [
            'family',
            'size',
            'care-level',
            'toxic',
            'light-requirements',
            'height',
        ].includes(attribute.name)
    );
    //get pairs of key value ["attribute name", value]
    const transformedAttributes: AttributesList = attributes.map(
        (attribute) => {
            let value: string | number | boolean | Sizes | Care | Light;
            if (
                typeof attribute.value === 'object' &&
                'label' in attribute.value
            ) {
                value = attribute.value.label;
            } else {
                value = attribute.value;
            }
            return [attribute.name, value];
        }
    );
    const attributesObject: CustomAttributes = Object.fromEntries(
        transformedAttributes
    );

    const discount = prices[0].discounted
        ? prices[0].discounted.value.centAmount / 100
        : null;
    return {
        id: id,
        sku: sku,
        price: prices[0].value.centAmount / 100,
        discount: discount,
        images: images,
        family: attributesObject.family,
        size: attributesObject.size,
        care: attributesObject['care-level'],
        toxic: attributesObject.toxic,
        light: attributesObject['light-requirements'],
        height: attributesObject.height,
    };
};

export const transformProductData = (productData: ProductResponse) => {
    const {
        id,
        masterData: {
            current: { masterVariant, variants, name, description },
        },
    } = productData;

    const product: CustomProduct = {
        id: id,
        name: name['en-US'],
        description: description['en-US'],
        masterVariant: prepareVariant(masterVariant),
        variants: variants.map((variant) => prepareVariant(variant)),
    };

    return product;
};
