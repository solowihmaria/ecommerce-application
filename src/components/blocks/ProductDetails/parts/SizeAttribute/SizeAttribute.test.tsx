import { render, screen } from '@testing-library/react';
import { Care, Light, Sizes } from '../../../../../api/product/product.types';
import type { CustomProduct } from '../../../../../api/product/product.types';
import { SizeAttribute } from './SizeAttribute';
import '@testing-library/jest-dom';

const product: CustomProduct = {
    id: '1676d70b-87b1-478c-a71b-bfedb3756ba7',
    name: 'Monstera Deliciosa',
    description:
        "The iconic Monstera Deliciosa, also known as the Swiss Cheese Plant, is beloved for its dramatic split leaves and easy-going nature. This tropical beauty develops its characteristic fenestrations (holes) as it matures, creating a stunning architectural presence in any room. Native to Central America, it's perfect for beginners and thrives in medium light conditions. Keep away from pets and children as it's mildly toxic when ingested.",
    masterVariant: {
        id: 1,
        sku: 'MONSTERA001S',
        price: 25,
        discount: 22.5,
        images: [
            {
                url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Monstera%20Deliciosa-2-miBul_l2.jpg',
                label: 'Monstera Deliciosa',
                dimensions: {
                    w: 1080,
                    h: 1080,
                },
            },
            {
                url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Monstera%20Deliciosa-3-adC1Gs9C.jpg',
                label: 'Monstera Deliciosa',
                dimensions: {
                    w: 1080,
                    h: 1080,
                },
            },
            {
                url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Monstera%20Deliciosa-4-YadbLJJ_.jpg',
                label: 'Monstera Deliciosa',
                dimensions: {
                    w: 1080,
                    h: 1080,
                },
            },
        ],
        family: 'Monstera (Deliciosa)',
        size: Sizes.small,
        care: Care.easy,
        toxic: true,
        light: Light.medium,
        height: 30,
    },
    variants: [
        {
            id: 2,
            sku: 'MONSTERA001M',
            price: 45,
            discount: 40.5,
            images: [
                {
                    url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/Monstera%20Deliciosa-w-NcF8G7UQ.jpg',
                    dimensions: {
                        w: 1200,
                        h: 1200,
                    },
                    label: 'image',
                },
            ],
            family: 'Monstera (Deliciosa)',
            size: Sizes.medium,
            care: Care.easy,
            toxic: true,
            light: Light.medium,
            height: 60,
        },
        {
            id: 3,
            sku: 'MONSTERA001L',
            price: 75,
            discount: 67.5,
            images: [
                {
                    url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/1646961625_2-idei-cl-YUC5vGX0.jpg',
                    label: 'Monstera Deliciosa-large',
                    dimensions: {
                        w: 3024,
                        h: 4032,
                    },
                },
                {
                    url: 'https://images.cdn.europe-west1.gcp.commercetools.com/6d0a8854-2112-4168-a8ea-ae8880496080/63871a909af0d3409749-iV6fJFu7.jpg',
                    label: 'Monstera Deliciosa-large',
                    dimensions: {
                        w: 1200,
                        h: 1200,
                    },
                },
            ],
            family: 'Monstera (Deliciosa)',
            size: Sizes.large,
            care: Care.easy,
            toxic: true,
            light: Light.medium,
            height: 120,
        },
    ],
};

describe('<SizeAttribute/> component', () => {
    test('Render SizeAttribute component', () => {
        const onClick = jest.fn();
        render(
            <SizeAttribute
                product={product}
                currentProductVariant={product.masterVariant}
                onSizeChange={onClick}
            />
        );

        const size = screen.getByTestId('size-attribute');

        expect(size).toBeInTheDocument();
        expect(size).toMatchSnapshot();
    });
});
