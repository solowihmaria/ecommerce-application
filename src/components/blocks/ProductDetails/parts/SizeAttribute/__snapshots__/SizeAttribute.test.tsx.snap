// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SizeAttribute/> component Render Input component - without any properties 1`] = `
<ul
  class="sizeContainer"
  data-testid="size-attribute"
>
  <li>
    <button
      class="button primary button sizeActive size"
    >
      Small
    </button>
  </li>
  <li>
    <button
      class="button primary button size"
    >
      Medium
    </button>
  </li>
  <li>
    <button
      class="button primary button size"
    >
      Large
    </button>
  </li>
</ul>
`;
